### PR TITLE
Restructure conditional access handling in the control flow graph

### PR DESCRIPTION
**Pull request by 333fred**
We had a bug in the CFG builder that caused regions for the conditional access temp to be unnecessarily extended past liveliness. To fix this, I've restructured the visit order to visit the instance only when actually accessed, which fixes this issue. I've also added some testing to demonstrate the handling around dropped NoneOperations in XML element accesses, as those need some special logic to ensure we aren't dropping nodes.

### CODE REVIEW

File: src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs
Comment (333fred) [inline src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs]: >if (((CSharp.Syntax.ConditionalAccessExpressionSyntax)syntax.Parent).Expression == syntax) [](start = 36, length = 90)

AlekseyTs said:

> I think we should see if we want to fix the builder instead. Looking at the added `NestedConditionalAccess` unit-test, the life-time of capture 0 is extended unnecessarily.
> In any case I do not think we should be treating it as long lived (i.e. something that is known to span multiple statements). So the relaxation should probably be applied elsewhere

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IConditionalAccessExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IConditionalAccessExpression.cs]: >Block[B1] - Block [](start = 8, length = 17)

AlekseyTs said:

> I think we actually should have a nested region here. it should enclose B1, B2 and own capture [0].

File: src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:353]: AlekseyTs said:

> Should we have an issue to track disable tests like that?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:353]: I don't think that's particularly useful, as I'm not currently thinking that we'll be attempting to renable them in the future.

---
In reply to: [571246466](https://github.com/dotnet/roslyn/pull/51032#discussion_r571246466) [](ancestors = 571246466)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @AlekseyTs for review. Like my previous porting prs, the original commit is prefixed with `Ported` and the new one is prefixed with `Fixup` for diffability.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @AlekseyTs for reviews.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ConditionalAccessOperationTracker.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ConditionalAccessOperationTracker.cs:15]: >public readonly ArrayBuilder<IOperation>? Operations; [](start = 12, length = 53)

Consider adding a comment what these operations are and how they are related to conditional access.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3313]: Note to self: Add modifier.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3219]: >testExpression = VisitConditionalAccessTestExpression(currentConditionalAccess.Operation); [](start = 20, length = 90)

It look like an initialization of  ```testExpression``` with anything other than ```currentConditionalAccess.Operation``` is unexpected. For example, we do the following test below ```operation.Operation == testExpression```.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3315]: >if (operation is not NoneOperation) [](start = 12, length = 35)

I think we should have a comment explaining rational behind this condition.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3352]: >_ = VisitConditionalAccessTestExpression(_currentConditionalAccessTracker.Operations.Pop()); [](start = 16, length = 92)

Is this code path reachable? It feels like similar check in ```VisitConditionalAccess``` should prevent us from getting here.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3315]: If this is just an optimization, I am not sure if it is worth it.

---
In reply to: [573251761](https://github.com/dotnet/roslyn/pull/51032#discussion_r573251761) [](ancestors = 573251761)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3352]: I had thought so too, but I added it after `ConditionalAccessFlow_13` proved it was needed. Here's the original operation tree for the code (note the last conditional access with 2 none operation children):

```
IBlockOperation (3 statements) (OperationKind.Block, Type: null, IsInvalid) (Syntax: 'Sub M(c As  ... End Sub')
  IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid) (Syntax: 'result = c?.X?.@a1?.<e1>')
    Expression: 
      ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Object, IsInvalid, IsImplicit) (Syntax: 'result = c?.X?.@a1?.<e1>')
        Left: 
          IParameterReferenceOperation: result (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'result')
        Right: 
          IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Object, IsInvalid, IsImplicit) (Syntax: 'c?.X?.@a1?.<e1>')
            Conversion: CommonConversion (Exists: False, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
            Operand: 
              IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: 'c?.X?.@a1?.<e1>')
                Operation: 
                  IParameterReferenceOperation: c (OperationKind.ParameterReference, Type: C) (Syntax: 'c')
                WhenNotNull: 
                  IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: '.X?.@a1?.<e1>')
                    Operation: 
                      IFieldReferenceOperation: C.X As System.Xml.Linq.XElement (OperationKind.FieldReference, Type: System.Xml.Linq.XElement) (Syntax: '.X')
                        Instance Receiver: 
                          IConditionalAccessInstanceOperation (OperationKind.ConditionalAccessInstance, Type: C, IsImplicit) (Syntax: 'c')
                    WhenNotNull: 
                      IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: '.@a1?.<e1>')
                        Operation: 
                          IOperation:  (OperationKind.None, Type: null) (Syntax: '.@a1')
                        WhenNotNull: 
                          IOperation:  (OperationKind.None, Type: null, IsInvalid) (Syntax: '.<e1>')
  ILabeledOperation (Label: exit) (OperationKind.Labeled, Type: null, IsImplicit) (Syntax: 'End Sub')
    Statement: 
      null
  IReturnOperation (OperationKind.Return, Type: null, IsImplicit) (Syntax: 'End Sub')
    ReturnedValue: 
      null
```

---
In reply to: [573257576](https://github.com/dotnet/roslyn/pull/51032#discussion_r573257576) [](ancestors = 573257576)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3315]: This was not an optimization. It was, however, a bug. So I've fixed that, and the updates to `ConditionalAccess_13` will reflect this.

---
In reply to: [573401448](https://github.com/dotnet/roslyn/pull/51032#discussion_r573401448) [](ancestors = 573401448,573251761)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3219]: The updates to `ConditionalAccess_11` and `ConditionalAccess_12` will reflect addressing this feedback.

---
In reply to: [573174786](https://github.com/dotnet/roslyn/pull/51032#discussion_r573174786) [](ancestors = 573174786)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [776337916](https://github.com/dotnet/roslyn/pull/51032#issuecomment-776337916) [](ancestors = 776337916)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3347]: It looks like there was a change in behavior reflected in ```ConditionalAccessFlow_13```, which was likely caused by the fix in ```IsConditionalAccessInstancePresentInChildren```. Do we still need this check here after the fix? Could you please elaborate why?

---
In reply to: [573954488](https://github.com/dotnet/roslyn/pull/51032#discussion_r573954488) [](ancestors = 573954488,573257576)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [777158691](https://github.com/dotnet/roslyn/pull/51032#issuecomment-777158691) [](ancestors = 777158691)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5), assuming CI is passing

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3204]: This re-assignment is no longer needed, right?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (chsienki) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3208]: NIT: consider moving this below the while loop, to make it clearer how the operations are being mutated. It took me a bit to figure out how the stack was actually getting built.

**Review by chsienki — APPROVED** [review]
:shipit:

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3208]: I can't. The loop needs this to be set up correctly in case it encounters a NoneOperation.

**Pull request merged**
